src/audio/mixin_mixout/mixin_mixout_generic.c

```diff
@@ -0,0 +1,403 @@
+	}
+}
+
+static void remap_mix_channel_s16(struct audio_stream __sparse_cache *sink,
+				  uint8_t sink_channel_index,  uint8_t sink_channel_count,
+				  uint32_t start_frame, uint32_t mixed_frames,
+				  const struct audio_stream __sparse_cache *source,
+				  uint8_t source_channel_index, uint8_t source_channel_count,
+				  uint32_t frame_count, uint16_t gain)
+{
+	int16_t *dst, *src;
+	uint32_t frames_to_mix, frames_to_copy, left_frames;
+	uint32_t n, nmax, frames, i;
+
+	/* audio_stream_wrap() is required and is done below in a loop */
+	dst = (int16_t *)sink->w_ptr + start_frame * sink_channel_count + sink_channel_index;
+	src = (int16_t *)source->r_ptr + source_channel_index;
+
+	assert(mixed_frames >= start_frame);
+	frames_to_mix = MIN(mixed_frames - start_frame, frame_count);
+	frames_to_copy = frame_count - frames_to_mix;
+
+	for (left_frames = frames_to_mix; left_frames > 0; left_frames -= frames) {
+		src = audio_stream_wrap(source, src);
+		dst = audio_stream_wrap(sink, dst);
+		/* calculate the remaining samples*/
+		nmax = audio_stream_samples_without_wrap_s16(source, src);
+		n = MIN(left_frames * source_channel_count, nmax);
+		nmax = audio_stream_samples_without_wrap_s16(sink, dst);
+		n = MIN(n, nmax);
+		/* frames is the processed frame count in this loop*/
+		frames = 0;
+		for (i = 0; i < n; i += source_channel_count) {
+			*dst = sat_int16((int32_t)*dst +
+			       q_mults_16x16(*src, gain, IPC4_MIXIN_GAIN_SHIFT));
+			src += source_channel_count;
+			dst += sink_channel_count;
+			frames++;
+		}
+	}
+
+	for (left_frames = frames_to_copy; left_frames > 0; left_frames -= frames) {
+		src = audio_stream_wrap(source, src);
+		dst = audio_stream_wrap(sink, dst);
+		nmax = audio_stream_samples_without_wrap_s16(source, src);
+		n = MIN(left_frames * source_channel_count, nmax);
+		nmax = audio_stream_samples_without_wrap_s16(sink, dst);
+		n = MIN(n, nmax);
+		frames = 0;
+		for (i = 0; i < n; i += source_channel_count) {
+			*dst = (int16_t)q_mults_16x16(*src, gain, IPC4_MIXIN_GAIN_SHIFT);
+			src += source_channel_count;
+			dst += sink_channel_count;
+			frames++;
+		}
+	}
+}
+
+static void mute_channel_s16(struct audio_stream __sparse_cache *stream, uint8_t channel_index,
+			     uint32_t start_frame, uint32_t mixed_frames, uint32_t frame_count)
+{
+	uint32_t skip_mixed_frames, n, left_frames, i, channel_count, frames;
+	int16_t *ptr;
+
+	assert(mixed_frames >= start_frame);
+	skip_mixed_frames = mixed_frames - start_frame;
+
+	if (frame_count <= skip_mixed_frames)
+		return;
+	frame_count -= skip_mixed_frames;
+	channel_count = stream->channels;
+	/* audio_stream_wrap() is needed here and it is just below in a loop */
+	ptr = (int16_t *)stream->w_ptr + mixed_frames * stream->channels + channel_index;
+
+	for (left_frames = frame_count; left_frames; left_frames -= frames) {
+		ptr = audio_stream_wrap(stream, ptr);
+		n = audio_stream_samples_without_wrap_s16(stream, ptr);
+		n =  MIN(left_frames * channel_count, n);
```

Comment:
I'm not putting equations to MIN() because of risk of becoming evaluated twice.

Reply:
ack, the compiler will likely optimise this, but better avoid this with our "trivial" definition of `MIN()`. More complex definitions avoid this by using local variables to evaluate arguments only once. Since we don't do that, better not use expressions in `MIN()` arguments. Also there's a superfluous space there.